WordPress/gutenberg
# Block Variations Transforms: Display as icon buttons if each block variation has a unique icon

## What?


Following on from a comment https://github.com/WordPress/gutenberg/pull/39710#issuecomment-1077417494 by @jasmussen, this PR looks at swapping out the Transform to Variation dropdown list in the block inspector for a list of buttons (similar to orientation and justification controls).

Because it's possible that there are blocks out in the wild that utilise variation transformations where each variation does not have a unique icon, I've preserved the dropdown list as a fallback in the case that each variation does not have a unique icon. 

## Why?


The Transform to Variation dropdown doesn't immediately signal to a user what they can do with each variation, hopefully switching to icons makes it a little clearer.

## How?


* Check that each variation has a unique icon, and if so, render a set of buttons for each variation.
* Remove internal state for the `__experimentalBlockVariationTransforms` component and switch to using the `getActiveBlockVariation` selector, as the current approach didn't appear to play well with the Group block's variations which involve an `isActive` check that is more tolerant of empty / default values than the existing checks (this fixes a bug where when a Group block was initially inserted, the Group button wasn't highlighted)

## Testing Instructions





1. Insert a Group block, and check that each of the transform buttons works correctly
2. From the Row block, switch the orientation to be vertical — this should update the block to be the Stack variation, and the Stack variation should be automatically selected in the list of buttons
3. Also test with another block that uses transforms, e.g. add a Next Post block and check that you can switch between that and the Previous post variation.
4. For thoroughness, in the Group block's `variations.js` file ([this line](https://github.com/wordpress/gutenberg/blob/8201df837eda8e8291d032951560710c6f564077/packages/block-library/src/group/variations.js#L17)), try commenting out the group icon. Reload the editor and check that the Transform to Variation dropdown menu works as before.

## Screenshots or screencast 

Screengrab of usage (note that switching orientation immediately updates the selected variation):

![block-variations-buttons](https://user-images.githubusercontent.com/14988353/161695156-beca7db8-1e9d-4c66-b88d-3a85de047f5f.gif)

Screenshots:

| Group block | Next post block |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/161694763-3ddc5564-2ecb-4fe5-bd38-c9716d0357af.png) | ![image](https://user-images.githubusercontent.com/14988353/161694665-c0b6baf4-dbae-4e70-98ca-cea87f0202cc.png) |
